include/boost/nowide/config.hpp

```diff
@@ -50,7 +50,7 @@
 
 //! @endcond
 
-#if(defined(__GNUC__) || defined(__MINGW32__) && !defined(__APPLE__))
+#if(defined(__GNUC__) && !defined(__APPLE__) || defined(__MINGW32__))
```

Comment:
To avoid ambiguous evaluation order you can use parantheses to group the sub-expressions too. Or as an even easier alternative: Move the `__APPLE__` branch to the top to avoid having to check for the !APPLE case.